scipy/stats/_qmc.py

```diff
@@ -0,0 +1,1493 @@
+"""Quasi-Monte Carlo engines and helpers."""
+from abc import ABC, abstractmethod
+import math
+import warnings
+
+import numpy as np
+
+from scipy.optimize import brute
+from scipy._lib._util import check_random_state
+from scipy.optimize import basinhopping
+from scipy.stats import norm
+from scipy.stats._sobol import (
+    initialize_v, _cscramble, _fill_p_cumulative, _draw, _fast_forward,
+    _categorize, initialize_direction_numbers, _MAXDIM, _MAXBIT
+)
+
+__all__ = ['scale', 'discrepancy', 'QMCEngine', 'Sobol', 'Halton',
+           'OrthogonalLatinHypercube', 'LatinHypercube', 'OptimalDesign',
+           'multinomial_qmc', 'NormalQMC', 'MultivariateNormalQMC']
+
+
+def scale(sample, bounds, reverse=False):
+    """Sample scaling from unit hypercube to bounds range.
+
+    To convert a sample from :math:`[0, 1)` to :math:`[a, b), b>a`, the
+    following transformation is used:
+
+    .. math::
+
+        (b - a) * sample + a
+
+    Parameters
+    ----------
+    sample : array_like (n_samples, dim)
+        Sample to scale.
+    bounds : tuple or array_like ([min, dim], [max, dim])
+        Desired range of transformed data. The transformation apply the bounds
+        on the sample and not the theoretical space, unit cube. Thus min and
+        max values of the sample will coincide with the bounds.
+    reverse : bool
+        Reverse the transformation, from bounds range to unit hypercube.
+
+    Returns
+    -------
+    sample : array_like (n_samples, dim)
+        Scaled-sample.
+
+    Examples
+    --------
+    >>> from scipy.stats import qmc
+    >>> bounds = [[-2, 0],
+    ...           [6, 5]]
+    >>> sample = [[0.5 , 0.5 ],
+    ...           [0.75, 0.25]]
+    >>> qmc.scale(sample, bounds)
+    array([[2.  , 2.5 ],
+           [4.  , 1.25]])
+
+    """
+    bounds = np.asarray(bounds)
+    min_ = np.min(bounds, axis=0)
+    max_ = np.max(bounds, axis=0)
+    if not reverse:
+        return sample * (max_ - min_) + min_
+    else:
+        return (sample - min_) / (max_ - min_)
+
+
+def discrepancy(sample, iterative=False, method='CD'):
+    """Discrepancy on a given sample.
+
+    Parameters
+    ----------
+    sample : array_like (n_samples, dim)
+        The sample to compute the discrepancy from.
+    iterative : bool
+        Must be False if not using it for updating the discrepancy.
```

Comment:
What does this mean?

Reply:
It means that if you set it to True, it will assume that you want to compute the discrepancy for N+1 points. So if you want 10 samples but right now you just have 9. You can compute the discrepancy like you would have 10. Then if you have a new sample, you can get the discrepancy using a cheaper method with `_update_discrepancy`. This is useful if you want to construct an iterative design and want to test some candidates.

Reply:
Maybe the two functions should be merged. But I separated them as the update is something quite specific and advanced and I did not want to have an interface too heavy here. Maybe a better option would be to optionally ask for the number of sample and default to the size of the sample passed.

Reply:
I still don't understand how to use this. Maybe provide an example, and then once I understand I can suggest ways this could be adjusted that would have helped me to understand.

Reply:
> I had an example but you told me to remove it because it was using a private method.

I see how it might look like my position had flip-flopped. I try to avoid that, but there is a chance that will happen in a PR this big, and I'm sorry if it does. Sometimes iteration is necessary to get to the final result.

I don't think that's happening here though. Here is where I'm coming from:

1. We do our best to completely document all features of public functions.
2. To my knowledge, we avoid having private functions (those beginning with an underscore) that are not used by public functions. 
3. Private functions that _are_ use by public functions are great, but we don't mention them in the documentation.

So if use of `discrepancy(..., iterative=True)` only makes sense together with use of `_update_discrepancy`, then you need to make a decision: either `_update_discrepancy` becomes public, or you remove the `iterative=True` option, or you change something else to satisfy rules 1-3. If `iterative=True` can be used sensibly without `_update_discrepancy`, then I don't think I've seen an example of that; please send one in the comments so I can better advise.

Does that help?

Reply:
As said in the other comment, I made `_update_discrepancy` public so I added an example in the docstring.